core/src/main/scala/shapeless/ops/hlists.scala

```diff
@@ -1142,6 +1142,130 @@ object hlist {
+          def apply(l: H :: T, m: M): Out = l.head :: u(l.tail, r(m)._2)
+        }
+  }
+
+  /**
+   * Type class supporting `HList` intersection. In case of duplicate types, this operation is a multiset intersection.
+   * If type `T` appears n times in this `HList` and m < n times in `M`, the resulting `HList` contains the first m
+   * elements of type `T` in this `HList`.
+   *
+   * Also available if `M` contains types absent in this `HList`.
+   *
+   * @author Olivier Blanvillain
+   */
+  trait Intersection[L <: HList, M <: HList] extends DepFn1[L] with Serializable { type Out <: HList }
+
+  trait LowPriorityIntersection {
+    type Aux[L <: HList, M <: HList, Out0 <: HList] = Intersection[L, M] { type Out = Out0 }
+
+    implicit def hlistIntersection1[H, T <: HList, M <: HList]
+      (implicit i: Intersection[T, M]): Aux[H :: T, M, i.Out] =
+        new Intersection[H :: T, M] {
+          type Out = i.Out
+          def apply(l: H :: T): Out = i(l.tail)
+        }
+  }
+
+  object Intersection extends LowPriorityIntersection {
+    def apply[L <: HList, M <: HList](implicit intersection: Intersection[L, M]): Aux[L, M, intersection.Out] = intersection
+
+    implicit def hnilIntersection[M <: HList]: Aux[HNil, M, HNil] =
+      new Intersection[HNil, M] {
+        type Out = HNil
+        def apply(l: HNil): Out = HNil
+      }
+
+    implicit def hlistIntersection2[H, T <: HList, M <: HList, MR <: HList]
+      (implicit
+        r: Remove.Aux[M, H, (H, MR)],
+        i: Intersection[T, MR]
+      ): Aux[H :: T, M, H :: i.Out] =
+        new Intersection[H :: T, M] {
+          type Out = H :: i.Out
+          def apply(l: H :: T): Out = l.head :: i(l.tail)
+        }
+  }
+
+  /**
+   * Type class supporting `HList` subtraction. In case of duplicate types, this operation is a multiset difference.
+   * If type `T` appears n times in this `HList` and m < n times in `M`, the resulting `HList` contains the last n - m
+   * elements of type `T` in this `HList`.
+   *
+   * Also available if `M` contains types absent in this `HList`.
+   *
+   * @author Olivier Blanvillain
+   */
+  trait Diff[L <: HList, M <: HList] extends DepFn1[L] with Serializable { type Out <: HList }
+
+  trait LowPriorityDiff {
+    type Aux[L <: HList, M <: HList, Out0] = Diff[L, M] { type Out = Out0 }
+
+    implicit def hconsDiff1[L <: HList, H, T <: HList]
+      (implicit d: Diff[L, T]): Aux[L, H :: T, d.Out] =
+        new Diff[L, H :: T] {
+          type Out = d.Out
+          def apply(l: L): Out = d(l)
+        }
+  }
+
+  object Diff extends LowPriorityDiff {
+    def apply[L <: HList, M <: HList](implicit diff: Diff[L, M]): Aux[L, M, diff.Out] = diff
+
+    implicit def hnilDiff[L <: HList]: Aux[L, HNil, L] =
+      new Diff[L, HNil] {
+        type Out = L
+        def apply(l: L): Out = l
+      }
+
+    implicit def hconsDiff2[L <: HList, LT <: HList, H, T <: HList]
+      (implicit
+        r: Remove.Aux[L, H, (H, LT)],
+        d: Diff[LT, T]
```

Comment:
Note: recusing on `Diff[LT, H :: T]` instead of `Diff[LT, T]` would work as well. With `Diff[LT, H :: T]`, all duplicate types `A` of `L` are removed if `A` is present in `M`, while with `Diff[LT, T]`, at most one type of `L` is removed for each type of `M`. The same consideration applies to `Intersection`.